### PR TITLE
Multiple head architecture for cuda backend

### PR DESCRIPTION
**Pull request by almaudoh**

### CODE REVIEW

File: src/neural/cuda/network_cudnn.cc [hunk 1]
```diff
@@ -950,7 +954,7 @@ class CudnnNetwork : public Network {
   // Apparently nvcc doesn't see constructor invocations through make_unique.
   // This function invokes constructor just to please complier and silence
   // warning. Is never called (but compiler thinks that it could).
-  void UglyFunctionToSilenceNvccWarning() { InputsOutputs io(0, false, false); }
+  void UglyFunctionToSilenceNvccWarning() { InputsOutputs io(0, false, false, false); }
 
  private:
   const NetworkCapabilities capabilities_;
```
Comment (borg323) [inline src/neural/cuda/network_cudnn.cc:960]: Later in the file, after line 1098: `case pblczero::NetworkFormat::NETWORK_ATTENTIONBODY_WITH_HEADFORMAT:` you should add the `case pblczero::NetworkFormat::NETWORK_ATTENTIONBODY_WITH_MULTIHEADFORMAT:` to switch to cuda.

File: src/neural/cuda/network_cuda.cc [hunk 2]
```diff
@@ -359,6 +356,22 @@ class CudaNetwork : public Network {
 
     ActivationFunction act = mish_net ? ACTIVATION_MISH : ACTIVATION_RELU;
 
+    std::string policy_head =
+        options.GetOrDefault<std::string>("policy_head", "vanilla");
+    // Check that selected policy head exists.
+    if (policy_head == "optimistic") policy_head = "optimistic_st";
```
Comment (borg323) [inline src/neural/cuda/network_cuda.cc:362]: Better rename the map key in `network_legacy.cc` to `optimistic`.

File: src/neural/cuda/layers.cc [hunk 3]
```diff
@@ -2073,12 +2104,13 @@ AttentionBody<DataType>::AttentionBody(const LegacyWeights& weights,
...
   int num_encoders = weights.encoder.size();
-  float alpha = (float)pow(2.0 * num_encoders, 0.25);
+  float alpha = (float)pow(2.0 * num_encoders, -0.25);
   for (const auto& enc : weights.encoder) {
     EncoderBlock<DataType>* pW = new EncoderBlock<DataType>(
         enc, scratch, encoder_head_count_, embedding_op_size_, alpha,
         smolgen_global_, smolgen_global_size_, max_batch_size,
-        activations_.smolgen_activation, activations_.ffn_activation);
+        activations_.smolgen_activation, activations_.ffn_activation,
+        new_encoding_ ? 1e-3 : 1e-6);
```
Comment (borg323) [inline src/neural/cuda/layers.cc:2113]: Should the 1e-6 be scaled by depth for true equivalence with the old code?

File: src/neural/cuda/network_cudnn.cc (hunk 1 above)
Comment (almaudoh) [reply src/neural/cuda/network_cudnn.cc:960]: Done

File: src/neural/cuda/layers.cc (hunk 3 above)
Comment (borg323) [reply src/neural/cuda/layers.cc:2113]: After doing some experiments, the scaling results in even lower epsilon values, so not worth the effort.

File: src/neural/cuda/network_cuda.cc [hunk 4]
```diff
@@ -49,24 +49,25 @@ using namespace cudnn_backend;
 template <typename DataType>
 class CudaNetwork;
 
-static size_t getMaxAttentionHeadSize(const LegacyWeights& weights, int N) {
-  const size_t embedding_op_size = weights.ip_pol_b.size();
-  const size_t policy_d_model = weights.ip2_pol_b.size();
-  assert(policy_d_model == weights.ip3_pol_b.size());
+static size_t getMaxAttentionHeadSize(const MultiHeadWeights& weights, int N) {
```
Comment (borg323) [inline src/neural/cuda/network_cuda.cc:52]: I think a better approach is to pass a `MultiHeadWeights::PolicyHead&` instead of `LegacyWeights&`.

Comment (almaudoh) [reply src/neural/cuda/network_cuda.cc]: Done

File: src/neural/cuda/inputs_outputs.h [hunk 5]
```diff
@@ -134,6 +136,8 @@ struct InputsOutputs {
 
   // cublas handle used to run the network
   cublasHandle_t cublas_;
+
+  bool has_moves_left_ = false;
```
Comment (mooskagh) [inline src/neural/cuda/inputs_outputs.h:140]: Wouldn't it be possible to just check op_moves_left_mem_ to be nullptr? (and initialize it like that)

Comment (almaudoh) [reply src/neural/cuda/inputs_outputs.h:140]: I did not find a very good way to test if the `op_moves_left_mem_` has been allocated. `op_moves_left_mem_ != nullptr` doesn't work as even the unallocated one still gets a pointer value. The only thing I have seen is that `op_moves_left_mem_ == op_moves_left_mem_gpu_` for cases where it has been allocated (and would be different if not allocated). Though I'm not sure if that is a safe or good way to go about it.

Comment (almaudoh) [reply src/neural/cuda/inputs_outputs.h]: I didn't understand the comment initially. Fixed now. Initialized to `nullptr` and then used as the check for deallocation.

**Review by borg323 — APPROVED** [review]

**Pull request merged**